<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>VCF Parsing &mdash; fastDFE 1.1.1 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" type="text/css" />
      <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="VCF Annotation" href="annotation.html" />
    <link rel="prev" title="Spectra class" href="spectra.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            fastDFE
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../reference/introduction.html">Introduction</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Python Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../reference/Python/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/Python/quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/Python/inference.html">DFE Inference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/Python/config.html">Configuration files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/Python/parametrizations.html">DFE parametrizations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/Python/spectra.html">Working with SFS</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/Python/parser.html">VCF Parsing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/Python/miscellaneous.html">Miscellaneous</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">R Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../reference/R/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/R/quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/R/inference.html">DFE Inference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/R/config.html">Configuration files</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/R/parametrizations.html">DFE parametrizations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/R/spectra.html">Working with SFS</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/R/parser.html">VCF Parsing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/R/miscellaneous.html">Miscellaneous</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Reference</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="inference.html">Inference</a></li>
<li class="toctree-l1"><a class="reference internal" href="config.html">Configuration class</a></li>
<li class="toctree-l1"><a class="reference internal" href="parametrization.html">DFE Parametrizations</a></li>
<li class="toctree-l1"><a class="reference internal" href="spectrum.html">Spectrum class</a></li>
<li class="toctree-l1"><a class="reference internal" href="spectra.html">Spectra class</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">VCF Parsing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#fastdfe.parser.Stratification"><code class="docutils literal notranslate"><span class="pre">Stratification</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Stratification.__init__"><code class="docutils literal notranslate"><span class="pre">Stratification.__init__()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Stratification.parser"><code class="docutils literal notranslate"><span class="pre">Stratification.parser</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Stratification.n_valid"><code class="docutils literal notranslate"><span class="pre">Stratification.n_valid</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Stratification.get_type"><code class="docutils literal notranslate"><span class="pre">Stratification.get_type()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Stratification.get_types"><code class="docutils literal notranslate"><span class="pre">Stratification.get_types()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#fastdfe.parser.SNPStratification"><code class="docutils literal notranslate"><span class="pre">SNPStratification</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#fastdfe.parser.BaseContextStratification"><code class="docutils literal notranslate"><span class="pre">BaseContextStratification</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.BaseContextStratification.__init__"><code class="docutils literal notranslate"><span class="pre">BaseContextStratification.__init__()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.BaseContextStratification.n_flanking"><code class="docutils literal notranslate"><span class="pre">BaseContextStratification.n_flanking</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.BaseContextStratification.contig"><code class="docutils literal notranslate"><span class="pre">BaseContextStratification.contig</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.BaseContextStratification.get_type"><code class="docutils literal notranslate"><span class="pre">BaseContextStratification.get_type()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.BaseContextStratification.get_types"><code class="docutils literal notranslate"><span class="pre">BaseContextStratification.get_types()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#fastdfe.parser.BaseTransitionStratification"><code class="docutils literal notranslate"><span class="pre">BaseTransitionStratification</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.BaseTransitionStratification.get_type"><code class="docutils literal notranslate"><span class="pre">BaseTransitionStratification.get_type()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.BaseTransitionStratification.get_types"><code class="docutils literal notranslate"><span class="pre">BaseTransitionStratification.get_types()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#fastdfe.parser.TransitionTransversionStratification"><code class="docutils literal notranslate"><span class="pre">TransitionTransversionStratification</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.TransitionTransversionStratification.get_type"><code class="docutils literal notranslate"><span class="pre">TransitionTransversionStratification.get_type()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.TransitionTransversionStratification.get_types"><code class="docutils literal notranslate"><span class="pre">TransitionTransversionStratification.get_types()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#fastdfe.parser.AncestralBaseStratification"><code class="docutils literal notranslate"><span class="pre">AncestralBaseStratification</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.AncestralBaseStratification.get_type"><code class="docutils literal notranslate"><span class="pre">AncestralBaseStratification.get_type()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.AncestralBaseStratification.get_types"><code class="docutils literal notranslate"><span class="pre">AncestralBaseStratification.get_types()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#fastdfe.parser.DegeneracyStratification"><code class="docutils literal notranslate"><span class="pre">DegeneracyStratification</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.DegeneracyStratification.__init__"><code class="docutils literal notranslate"><span class="pre">DegeneracyStratification.__init__()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.DegeneracyStratification.get_degeneracy"><code class="docutils literal notranslate"><span class="pre">DegeneracyStratification.get_degeneracy</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.DegeneracyStratification.get_type"><code class="docutils literal notranslate"><span class="pre">DegeneracyStratification.get_type()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.DegeneracyStratification.get_types"><code class="docutils literal notranslate"><span class="pre">DegeneracyStratification.get_types()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#fastdfe.parser.SynonymyStratification"><code class="docutils literal notranslate"><span class="pre">SynonymyStratification</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.SynonymyStratification.get_types"><code class="docutils literal notranslate"><span class="pre">SynonymyStratification.get_types()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.SynonymyStratification.get_type"><code class="docutils literal notranslate"><span class="pre">SynonymyStratification.get_type()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#fastdfe.parser.VEPStratification"><code class="docutils literal notranslate"><span class="pre">VEPStratification</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.VEPStratification.info_tag"><code class="docutils literal notranslate"><span class="pre">VEPStratification.info_tag</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.VEPStratification.get_types"><code class="docutils literal notranslate"><span class="pre">VEPStratification.get_types()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.VEPStratification.get_type"><code class="docutils literal notranslate"><span class="pre">VEPStratification.get_type()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#fastdfe.parser.SnpEffStratification"><code class="docutils literal notranslate"><span class="pre">SnpEffStratification</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.SnpEffStratification.info_tag"><code class="docutils literal notranslate"><span class="pre">SnpEffStratification.info_tag</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#fastdfe.parser.GenomePositionDependentStratification"><code class="docutils literal notranslate"><span class="pre">GenomePositionDependentStratification</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#fastdfe.parser.ContigStratification"><code class="docutils literal notranslate"><span class="pre">ContigStratification</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.ContigStratification.get_type"><code class="docutils literal notranslate"><span class="pre">ContigStratification.get_type()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.ContigStratification.get_types"><code class="docutils literal notranslate"><span class="pre">ContigStratification.get_types()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#fastdfe.parser.ChunkedStratification"><code class="docutils literal notranslate"><span class="pre">ChunkedStratification</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.ChunkedStratification.__init__"><code class="docutils literal notranslate"><span class="pre">ChunkedStratification.__init__()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.ChunkedStratification.n_chunks"><code class="docutils literal notranslate"><span class="pre">ChunkedStratification.n_chunks</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.ChunkedStratification.chunk_sizes"><code class="docutils literal notranslate"><span class="pre">ChunkedStratification.chunk_sizes</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.ChunkedStratification.counter"><code class="docutils literal notranslate"><span class="pre">ChunkedStratification.counter</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.ChunkedStratification.get_types"><code class="docutils literal notranslate"><span class="pre">ChunkedStratification.get_types()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.ChunkedStratification.get_type"><code class="docutils literal notranslate"><span class="pre">ChunkedStratification.get_type()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#fastdfe.parser.TargetSiteCounter"><code class="docutils literal notranslate"><span class="pre">TargetSiteCounter</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.TargetSiteCounter.__init__"><code class="docutils literal notranslate"><span class="pre">TargetSiteCounter.__init__()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.TargetSiteCounter.n_target_sites"><code class="docutils literal notranslate"><span class="pre">TargetSiteCounter.n_target_sites</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.TargetSiteCounter.n_samples"><code class="docutils literal notranslate"><span class="pre">TargetSiteCounter.n_samples</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.TargetSiteCounter.count"><code class="docutils literal notranslate"><span class="pre">TargetSiteCounter.count()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#fastdfe.parser.Parser"><code class="docutils literal notranslate"><span class="pre">Parser</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Parser.__init__"><code class="docutils literal notranslate"><span class="pre">Parser.__init__()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Parser.target_site_counter"><code class="docutils literal notranslate"><span class="pre">Parser.target_site_counter</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Parser.n"><code class="docutils literal notranslate"><span class="pre">Parser.n</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Parser.include_samples"><code class="docutils literal notranslate"><span class="pre">Parser.include_samples</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Parser.exclude_samples"><code class="docutils literal notranslate"><span class="pre">Parser.exclude_samples</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Parser.skip_non_polarized"><code class="docutils literal notranslate"><span class="pre">Parser.skip_non_polarized</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Parser.stratifications"><code class="docutils literal notranslate"><span class="pre">Parser.stratifications</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Parser.annotations"><code class="docutils literal notranslate"><span class="pre">Parser.annotations</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Parser.filtrations"><code class="docutils literal notranslate"><span class="pre">Parser.filtrations</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Parser.n_skipped"><code class="docutils literal notranslate"><span class="pre">Parser.n_skipped</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Parser.n_no_ancestral"><code class="docutils literal notranslate"><span class="pre">Parser.n_no_ancestral</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Parser.sfs"><code class="docutils literal notranslate"><span class="pre">Parser.sfs</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Parser.subsample_mode"><code class="docutils literal notranslate"><span class="pre">Parser.subsample_mode</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastdfe.parser.Parser.parse"><code class="docutils literal notranslate"><span class="pre">Parser.parse()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="annotation.html">VCF Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="filtration.html">VCF Filtration</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">fastDFE</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">VCF Parsing</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/modules/parser.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-fastdfe.parser">
<span id="vcf-parsing"></span><span id="modules-parser"></span><h1>VCF Parsing<a class="headerlink" href="#module-fastdfe.parser" title="Permalink to this heading"></a></h1>
<p>A VCF parser that can be used to extract the site frequency spectrum (SFS) from a VCF file.
Stratifying the SFS is supported by providing a list of <a class="reference internal" href="#fastdfe.parser.Stratification" title="fastdfe.parser.Stratification"><code class="xref py py-class docutils literal notranslate"><span class="pre">Stratification</span></code></a> instances.</p>
<dl class="py class">
<dt class="sig sig-object py" id="fastdfe.parser.Stratification">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Stratification</span></span><a class="headerlink" href="#fastdfe.parser.Stratification" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">ABC</span></code></p>
<p>Abstract class for Stratifying the SFS by determining a site’s type based on its properties.</p>
<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.Stratification.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.Stratification.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Create instance.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.Stratification.parser">
<span class="sig-name descname"><span class="pre">parser</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code><span class="pre">[</span><a class="reference internal" href="#fastdfe.parser.Parser" title="fastdfe.parser.Parser"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parser</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#fastdfe.parser.Stratification.parser" title="Permalink to this definition"></a></dt>
<dd><p>Parser instance</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.Stratification.n_valid">
<span class="sig-name descname"><span class="pre">n_valid</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></em><a class="headerlink" href="#fastdfe.parser.Stratification.n_valid" title="Permalink to this definition"></a></dt>
<dd><p>The number of sites that didn’t have a type.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.Stratification.get_type">
<em class="property"><span class="pre">abstract</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cyvcf2.cyvcf2.Variant</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">fastdfe.io_handlers.DummyVariant</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.Stratification.get_type" title="Permalink to this definition"></a></dt>
<dd><p>Get type of given Variant. Only the types
given by <a class="reference internal" href="#fastdfe.parser.Stratification.get_types" title="fastdfe.parser.Stratification.get_types"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_types()</span></code></a> are valid, or <code class="docutils literal notranslate"><span class="pre">None</span></code> if
no type could be determined.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>variant</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code> | <code class="xref py py-class docutils literal notranslate"><span class="pre">DummyVariant</span></code>) – The vcf site</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Type of the variant</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.Stratification.get_types">
<em class="property"><span class="pre">abstract</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_types</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.Stratification.get_types" title="Permalink to this definition"></a></dt>
<dd><p>Get all possible types.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of types</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fastdfe.parser.SNPStratification">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SNPStratification</span></span><a class="headerlink" href="#fastdfe.parser.SNPStratification" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fastdfe.parser.Stratification" title="fastdfe.parser.Stratification"><code class="xref py py-class docutils literal notranslate"><span class="pre">Stratification</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">ABC</span></code></p>
<p>Abstract class for stratifications that can only handle SNPs. We need to issue a warning in this case.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fastdfe.parser.BaseContextStratification">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">BaseContextStratification</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fasta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_flanking</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aliases</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.BaseContextStratification" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fastdfe.parser.Stratification" title="fastdfe.parser.Stratification"><code class="xref py py-class docutils literal notranslate"><span class="pre">Stratification</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">FASTAHandler</span></code></p>
<p>Stratify the SFS by the base context of the mutation. The number of flanking bases
can be configured. Note that we attempt to take the ancestral allele as the
middle base. If <code class="docutils literal notranslate"><span class="pre">skip_non_polarized</span></code> is set to <code class="docutils literal notranslate"><span class="pre">False</span></code>, we use the reference
allele as the middle base.</p>
<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.BaseContextStratification.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fasta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_flanking</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aliases</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.BaseContextStratification.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Create instance. Note that we require a fasta file to be specified
for base context to be able to be inferred</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fasta</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – The fasta file path, possibly gzipped or a URL</p></li>
<li><p><strong>n_flanking</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The number of flanking bases</p></li>
<li><p><strong>aliases</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – Dictionary of aliases for the contigs in the VCF file, e.g. <code class="docutils literal notranslate"><span class="pre">{'chr1':</span> <span class="pre">['1']}</span></code>.
This is used to match the contig names in the VCF file with the contig names in the FASTA file and GFF file.</p></li>
<li><p><strong>cache</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to cache files that are downloaded from URLs</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.BaseContextStratification.n_flanking">
<span class="sig-name descname"><span class="pre">n_flanking</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></em><a class="headerlink" href="#fastdfe.parser.BaseContextStratification.n_flanking" title="Permalink to this definition"></a></dt>
<dd><p>The number of flanking bases</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.BaseContextStratification.contig">
<span class="sig-name descname"><span class="pre">contig</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code><span class="pre">[</span><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqRecord</span></code><span class="pre">]</span></em><a class="headerlink" href="#fastdfe.parser.BaseContextStratification.contig" title="Permalink to this definition"></a></dt>
<dd><p>The current contig</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.BaseContextStratification.get_type">
<span class="sig-name descname"><span class="pre">get_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cyvcf2.cyvcf2.Variant</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">fastdfe.io_handlers.DummyVariant</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.BaseContextStratification.get_type" title="Permalink to this definition"></a></dt>
<dd><p>Get the base context for a given mutation</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>variant</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code> | <code class="xref py py-class docutils literal notranslate"><span class="pre">DummyVariant</span></code>) – The vcf site</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Base context of the mutation</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.BaseContextStratification.get_types">
<span class="sig-name descname"><span class="pre">get_types</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.BaseContextStratification.get_types" title="Permalink to this definition"></a></dt>
<dd><p>Create all possible base contexts.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of contexts</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fastdfe.parser.BaseTransitionStratification">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">BaseTransitionStratification</span></span><a class="headerlink" href="#fastdfe.parser.BaseTransitionStratification" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fastdfe.parser.SNPStratification" title="fastdfe.parser.SNPStratification"><code class="xref py py-class docutils literal notranslate"><span class="pre">SNPStratification</span></code></a></p>
<p>Stratify the SFS by the base transition of the mutation, i.e., <code class="docutils literal notranslate"><span class="pre">A&gt;T</span></code>.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This stratification only works for SNPs. You thus need to update the number of mono-allelic sites manually.</p>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.BaseTransitionStratification.get_type">
<span class="sig-name descname"><span class="pre">get_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cyvcf2.cyvcf2.Variant</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">fastdfe.io_handlers.DummyVariant</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.BaseTransitionStratification.get_type" title="Permalink to this definition"></a></dt>
<dd><p>Get the base transition for the given variant.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>variant</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code> | <code class="xref py py-class docutils literal notranslate"><span class="pre">DummyVariant</span></code>) – The vcf site</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Base transition</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>NoTypeException</strong> – if not type could be determined</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.BaseTransitionStratification.get_types">
<span class="sig-name descname"><span class="pre">get_types</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.BaseTransitionStratification.get_types" title="Permalink to this definition"></a></dt>
<dd><p>Get all possible base transitions.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of contexts</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fastdfe.parser.TransitionTransversionStratification">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">TransitionTransversionStratification</span></span><a class="headerlink" href="#fastdfe.parser.TransitionTransversionStratification" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fastdfe.parser.BaseTransitionStratification" title="fastdfe.parser.BaseTransitionStratification"><code class="xref py py-class docutils literal notranslate"><span class="pre">BaseTransitionStratification</span></code></a></p>
<p>Stratify the SFS by whether we have a transition or transversion.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This stratification only works for SNPs. You thus need to update the number of mono-allelic sites manually.</p>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.TransitionTransversionStratification.get_type">
<span class="sig-name descname"><span class="pre">get_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cyvcf2.cyvcf2.Variant</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">fastdfe.io_handlers.DummyVariant</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.TransitionTransversionStratification.get_type" title="Permalink to this definition"></a></dt>
<dd><p>Get the mutation type (transition or transversion) for a given mutation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>variant</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code> | <code class="xref py py-class docutils literal notranslate"><span class="pre">DummyVariant</span></code>) – The vcf site</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Mutation type</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.TransitionTransversionStratification.get_types">
<span class="sig-name descname"><span class="pre">get_types</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.TransitionTransversionStratification.get_types" title="Permalink to this definition"></a></dt>
<dd><p>All possible mutation types (transition and transversion).</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of mutation types</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fastdfe.parser.AncestralBaseStratification">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">AncestralBaseStratification</span></span><a class="headerlink" href="#fastdfe.parser.AncestralBaseStratification" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fastdfe.parser.Stratification" title="fastdfe.parser.Stratification"><code class="xref py py-class docutils literal notranslate"><span class="pre">Stratification</span></code></a></p>
<p>Stratify the SFS by the base context of the mutation: the reference base.
If <code class="docutils literal notranslate"><span class="pre">skip_non_polarized</span></code> is set to <code class="docutils literal notranslate"><span class="pre">False</span></code>, we use the reference allele as
ancestral base. By default, we use the <code class="docutils literal notranslate"><span class="pre">AA</span></code> tag to determine the ancestral allele.</p>
<p>Any subclass of <code class="xref py py-class docutils literal notranslate"><span class="pre">AncestralAnnotation</span></code> can be used to annotate the ancestral allele.</p>
<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.AncestralBaseStratification.get_type">
<span class="sig-name descname"><span class="pre">get_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cyvcf2.cyvcf2.Variant</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">fastdfe.io_handlers.DummyVariant</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.AncestralBaseStratification.get_type" title="Permalink to this definition"></a></dt>
<dd><p>Get the type which is the reference allele.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>variant</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code> | <code class="xref py py-class docutils literal notranslate"><span class="pre">DummyVariant</span></code>) – The vcf site</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>reference allele</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.AncestralBaseStratification.get_types">
<span class="sig-name descname"><span class="pre">get_types</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.AncestralBaseStratification.get_types" title="Permalink to this definition"></a></dt>
<dd><p>The possible base types.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of contexts</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fastdfe.parser.DegeneracyStratification">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">DegeneracyStratification</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">custom_callback</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Callable</span><span class="p"><span class="pre">[</span></span><span class="p"><span class="pre">[</span></span><span class="pre">cyvcf2.cyvcf2.Variant</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.DegeneracyStratification" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fastdfe.parser.Stratification" title="fastdfe.parser.Stratification"><code class="xref py py-class docutils literal notranslate"><span class="pre">Stratification</span></code></a></p>
<p>Stratify SFS by degeneracy. We only consider sides which 4-fold degenerate (neutral) or
0-fold degenerate (selected) which facilitates counting.</p>
<p><a class="reference internal" href="annotation.html#fastdfe.annotation.DegeneracyAnnotation" title="fastdfe.annotation.DegeneracyAnnotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">DegeneracyAnnotation</span></code></a> can be used to annotate the degeneracy of a site.</p>
<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.DegeneracyStratification.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">custom_callback</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Callable</span><span class="p"><span class="pre">[</span></span><span class="p"><span class="pre">[</span></span><span class="pre">cyvcf2.cyvcf2.Variant</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.DegeneracyStratification.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the stratification.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>custom_callback</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Callable</span></code>[[<code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code>], <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – Custom callback to determine the type of mutation</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.DegeneracyStratification.get_degeneracy">
<span class="sig-name descname"><span class="pre">get_degeneracy</span></span><a class="headerlink" href="#fastdfe.parser.DegeneracyStratification.get_degeneracy" title="Permalink to this definition"></a></dt>
<dd><p>Custom callback to determine the degeneracy of mutation</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.DegeneracyStratification.get_type">
<span class="sig-name descname"><span class="pre">get_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cyvcf2.cyvcf2.Variant</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">fastdfe.io_handlers.DummyVariant</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.DegeneracyStratification.get_type" title="Permalink to this definition"></a></dt>
<dd><p>Get the degeneracy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>variant</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code> | <code class="xref py py-class docutils literal notranslate"><span class="pre">DummyVariant</span></code>) – The vcf site</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code>[<code class="docutils literal notranslate"><span class="pre">'neutral'</span></code>, <code class="docutils literal notranslate"><span class="pre">'selected'</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Type of the mutation</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>NoTypeException</strong> – If the mutation is not synonymous or non-synonymous</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.DegeneracyStratification.get_types">
<span class="sig-name descname"><span class="pre">get_types</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.DegeneracyStratification.get_types" title="Permalink to this definition"></a></dt>
<dd><p>Get all possible degeneracy type (<code class="docutils literal notranslate"><span class="pre">neutral</span></code> and <code class="docutils literal notranslate"><span class="pre">selected</span></code>).</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of contexts</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fastdfe.parser.SynonymyStratification">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SynonymyStratification</span></span><a class="headerlink" href="#fastdfe.parser.SynonymyStratification" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fastdfe.parser.SNPStratification" title="fastdfe.parser.SNPStratification"><code class="xref py py-class docutils literal notranslate"><span class="pre">SNPStratification</span></code></a></p>
<p>Stratify SFS by synonymy (neutral or selected).</p>
<p><a class="reference internal" href="annotation.html#fastdfe.annotation.SynonymyAnnotation" title="fastdfe.annotation.SynonymyAnnotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">SynonymyAnnotation</span></code></a> can be used to annotate the synonymy of a site.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This stratification only works for SNPs. You thus need to update the number of mono-allelic sites manually.</p>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.SynonymyStratification.get_types">
<span class="sig-name descname"><span class="pre">get_types</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.SynonymyStratification.get_types" title="Permalink to this definition"></a></dt>
<dd><p>Get all possible synonymy types (<code class="docutils literal notranslate"><span class="pre">neutral</span></code> and <code class="docutils literal notranslate"><span class="pre">selected</span></code>).</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of contexts</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.SynonymyStratification.get_type">
<span class="sig-name descname"><span class="pre">get_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cyvcf2.cyvcf2.Variant</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">fastdfe.io_handlers.DummyVariant</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.SynonymyStratification.get_type" title="Permalink to this definition"></a></dt>
<dd><p>Get the synonymy using the custom synonymy annotation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>variant</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code> | <code class="xref py py-class docutils literal notranslate"><span class="pre">DummyVariant</span></code>) – The vcf site</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code>[<code class="docutils literal notranslate"><span class="pre">'neutral'</span></code>, <code class="docutils literal notranslate"><span class="pre">'selected'</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Type of the mutation, either <code class="docutils literal notranslate"><span class="pre">neutral</span></code> or <code class="docutils literal notranslate"><span class="pre">selected</span></code></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fastdfe.parser.VEPStratification">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">VEPStratification</span></span><a class="headerlink" href="#fastdfe.parser.VEPStratification" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fastdfe.parser.SynonymyStratification" title="fastdfe.parser.SynonymyStratification"><code class="xref py py-class docutils literal notranslate"><span class="pre">SynonymyStratification</span></code></a></p>
<p>Stratify SFS by synonymy (neutral or selected) based on annotation provided by VEP.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This stratification only works for SNPs. You thus need to update the number of mono-allelic sites manually.</p>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.VEPStratification.info_tag">
<span class="sig-name descname"><span class="pre">info_tag</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'CSQ'</span></em><a class="headerlink" href="#fastdfe.parser.VEPStratification.info_tag" title="Permalink to this definition"></a></dt>
<dd><p>The tag used by VEP to annotate the synonymy</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.VEPStratification.get_types">
<span class="sig-name descname"><span class="pre">get_types</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.VEPStratification.get_types" title="Permalink to this definition"></a></dt>
<dd><p>Get all possible synonymy types (<code class="docutils literal notranslate"><span class="pre">neutral</span></code> and <code class="docutils literal notranslate"><span class="pre">selected</span></code>).</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of contexts</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.VEPStratification.get_type">
<span class="sig-name descname"><span class="pre">get_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cyvcf2.cyvcf2.Variant</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">fastdfe.io_handlers.DummyVariant</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.VEPStratification.get_type" title="Permalink to this definition"></a></dt>
<dd><p>Get the synonymy of a site.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>variant</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code> | <code class="xref py py-class docutils literal notranslate"><span class="pre">DummyVariant</span></code>) – The vcf site</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code>[<code class="docutils literal notranslate"><span class="pre">'neutral'</span></code>, <code class="docutils literal notranslate"><span class="pre">'selected'</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Type of the mutation, either <code class="docutils literal notranslate"><span class="pre">neutral</span></code> or <code class="docutils literal notranslate"><span class="pre">selected</span></code></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fastdfe.parser.SnpEffStratification">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SnpEffStratification</span></span><a class="headerlink" href="#fastdfe.parser.SnpEffStratification" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fastdfe.parser.VEPStratification" title="fastdfe.parser.VEPStratification"><code class="xref py py-class docutils literal notranslate"><span class="pre">VEPStratification</span></code></a></p>
<p>Stratify SFS by synonymy (neutral or selected) based on annotation provided by SnpEff.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This stratification only works for SNPs. You thus need to update the number of mono-allelic sites manually.</p>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.SnpEffStratification.info_tag">
<span class="sig-name descname"><span class="pre">info_tag</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'ANN'</span></em><a class="headerlink" href="#fastdfe.parser.SnpEffStratification.info_tag" title="Permalink to this definition"></a></dt>
<dd><p>The tag used by SnpEff to annotate the synonymy</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fastdfe.parser.GenomePositionDependentStratification">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">GenomePositionDependentStratification</span></span><a class="headerlink" href="#fastdfe.parser.GenomePositionDependentStratification" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fastdfe.parser.Stratification" title="fastdfe.parser.Stratification"><code class="xref py py-class docutils literal notranslate"><span class="pre">Stratification</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">ABC</span></code></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fastdfe.parser.ContigStratification">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ContigStratification</span></span><a class="headerlink" href="#fastdfe.parser.ContigStratification" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fastdfe.parser.GenomePositionDependentStratification" title="fastdfe.parser.GenomePositionDependentStratification"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomePositionDependentStratification</span></code></a></p>
<p>Stratify SFS by contig.</p>
<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.ContigStratification.get_type">
<span class="sig-name descname"><span class="pre">get_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cyvcf2.cyvcf2.Variant</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">fastdfe.io_handlers.DummyVariant</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.ContigStratification.get_type" title="Permalink to this definition"></a></dt>
<dd><p>Get the contig.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>variant</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code> | <code class="xref py py-class docutils literal notranslate"><span class="pre">DummyVariant</span></code>) – The vcf site</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The contig name</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.ContigStratification.get_types">
<span class="sig-name descname"><span class="pre">get_types</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.ContigStratification.get_types" title="Permalink to this definition"></a></dt>
<dd><p>Get all possible contig type.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of contexts</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fastdfe.parser.ChunkedStratification">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ChunkedStratification</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_chunks</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.ChunkedStratification" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fastdfe.parser.GenomePositionDependentStratification" title="fastdfe.parser.GenomePositionDependentStratification"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomePositionDependentStratification</span></code></a></p>
<p>Stratify SFS by creating <code class="docutils literal notranslate"><span class="pre">n</span></code> contiguous chunks of roughly equal size.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Since the total number of sites is not known in advance, we cannot create contiguous
chunks of exactly equal size.</p>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.ChunkedStratification.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_chunks</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.ChunkedStratification.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the stratification.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n_chunks</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Number of sites per window</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.ChunkedStratification.n_chunks">
<span class="sig-name descname"><span class="pre">n_chunks</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></em><a class="headerlink" href="#fastdfe.parser.ChunkedStratification.n_chunks" title="Permalink to this definition"></a></dt>
<dd><p>Number of chunks</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.ChunkedStratification.chunk_sizes">
<span class="sig-name descname"><span class="pre">chunk_sizes</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code><span class="pre">[</span><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code><span class="pre">[</span><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code><span class="pre">]]</span></em><a class="headerlink" href="#fastdfe.parser.ChunkedStratification.chunk_sizes" title="Permalink to this definition"></a></dt>
<dd><p>List of chunk sizes</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.ChunkedStratification.counter">
<span class="sig-name descname"><span class="pre">counter</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></em><a class="headerlink" href="#fastdfe.parser.ChunkedStratification.counter" title="Permalink to this definition"></a></dt>
<dd><p>Number of sites seen so far</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.ChunkedStratification.get_types">
<span class="sig-name descname"><span class="pre">get_types</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.ChunkedStratification.get_types" title="Permalink to this definition"></a></dt>
<dd><p>Get all possible window types.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of contexts</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.ChunkedStratification.get_type">
<span class="sig-name descname"><span class="pre">get_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cyvcf2.cyvcf2.Variant</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">fastdfe.io_handlers.DummyVariant</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.ChunkedStratification.get_type" title="Permalink to this definition"></a></dt>
<dd><p>Get the type.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>variant</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code> | <code class="xref py py-class docutils literal notranslate"><span class="pre">DummyVariant</span></code>) – The vcf site</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The type</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fastdfe.parser.TargetSiteCounter">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">TargetSiteCounter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_target_sites</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_samples</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1000000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.TargetSiteCounter" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class for counting the number of target sites when parsing a VCF file that does not contain monomorphic sites.
This class is used in conjunction with <a class="reference internal" href="#fastdfe.parser.Parser" title="fastdfe.parser.Parser"><code class="xref py py-class docutils literal notranslate"><span class="pre">Parser</span></code></a> and samples sites from the given fasta
file that are found in between variants on the same contig that were parsed in the VCF.
Ideally, we obtain the SFS by parsing VCF files that contain both mono- and polymorphic sites. This is because
we need to know about the number of mutational opportunities for synonymous and non-synonymous sites which
contain plenty of information on the strength of selection. It is recommended to use a SNPFiltration when
using this class to avoid biasing the result by monomorphic sites present in the VCF file.</p>
<p>Note that we sample sites randomly to make sure we cover the entire genome when only a subset of the genome is
considered. This provides more speed as we don’t have to parse the entire FASTA file. The actual number of
monomorphic sites is then extrapolated from <code class="docutils literal notranslate"><span class="pre">n_target_sites</span></code>.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This class is not compatible with stratifications based on info tags that are pre-defined in the VCF file, as
opposed to those added dynamically using the <code class="docutils literal notranslate"><span class="pre">annotations</span></code> argument of the parser. We also need to
stratify mono-allelic sites which, in this case, won’t be present in the VCF file so that they have no
info tags when sampling from the FASTA file, and are thus ignored by the stratifications. However, using the
<code class="docutils literal notranslate"><span class="pre">annotations</span></code> argument of the parser, the info tags the stratifications are based on are added on-the-fly,
also for monomorphic sites sampled from the FASTA file.</p>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.TargetSiteCounter.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_target_sites</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_samples</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1000000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.TargetSiteCounter.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize counter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n_target_sites</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The total number of sites (mono- and polymorphic) that would be present in the VCF file
if it contained monomorphic sites. This number should be considerably larger than the number of polymorphic
sites in the VCF file. This value is not extremely important for the DFE inference, the ratio of synonymous
to non-synonymous sites being more informative, but the order of magnitude should be correct, in any case.</p></li>
<li><p><strong>n_samples</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The number of sites to sample from the fasta file. Many sampled sites will not be valid as
they are non-coding. To obtain good estimates, a few thousand sites should be sampled per type of site
(depending on the stratifications used).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.TargetSiteCounter.n_target_sites">
<span class="sig-name descname"><span class="pre">n_target_sites</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code><span class="pre">[</span><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code><span class="pre">]</span></em><a class="headerlink" href="#fastdfe.parser.TargetSiteCounter.n_target_sites" title="Permalink to this definition"></a></dt>
<dd><p>The total number of sites considered when parsing the VCF</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.TargetSiteCounter.n_samples">
<span class="sig-name descname"><span class="pre">n_samples</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></em><a class="headerlink" href="#fastdfe.parser.TargetSiteCounter.n_samples" title="Permalink to this definition"></a></dt>
<dd><p>Number of samples</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.TargetSiteCounter.count">
<span class="sig-name descname"><span class="pre">count</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.TargetSiteCounter.count" title="Permalink to this definition"></a></dt>
<dd><p>Count the number of target sites.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of target sites</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fastdfe.parser.Parser">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Parser</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">vcf:</span> <span class="pre">str,</span> <span class="pre">n:</span> <span class="pre">int,</span> <span class="pre">gff:</span> <span class="pre">typing.Optional[str]</span> <span class="pre">=</span> <span class="pre">None,</span> <span class="pre">fasta:</span> <span class="pre">typing.Optional[str]</span> <span class="pre">=</span> <span class="pre">None,</span> <span class="pre">info_ancestral:</span> <span class="pre">str</span> <span class="pre">=</span> <span class="pre">'AA',</span> <span class="pre">skip_non_polarized:</span> <span class="pre">bool</span> <span class="pre">=</span> <span class="pre">True,</span> <span class="pre">stratifications:</span> <span class="pre">typing.List[fastdfe.parser.Stratification]</span> <span class="pre">=</span> <span class="pre">[],</span> <span class="pre">annotations:</span> <span class="pre">typing.List[fastdfe.annotation.Annotation]</span> <span class="pre">=</span> <span class="pre">[],</span> <span class="pre">filtrations:</span> <span class="pre">typing.List[fastdfe.filtration.Filtration]</span> <span class="pre">=</span> <span class="pre">[&lt;fastdfe.filtration.PolyAllelicFiltration</span> <span class="pre">object&gt;],</span> <span class="pre">include_samples:</span> <span class="pre">typing.Optional[typing.List[str]]</span> <span class="pre">=</span> <span class="pre">None,</span> <span class="pre">exclude_samples:</span> <span class="pre">typing.Optional[typing.List[str]]</span> <span class="pre">=</span> <span class="pre">None,</span> <span class="pre">max_sites:</span> <span class="pre">int</span> <span class="pre">=</span> <span class="pre">inf,</span> <span class="pre">seed:</span> <span class="pre">int</span> <span class="pre">|</span> <span class="pre">None</span> <span class="pre">=</span> <span class="pre">0,</span> <span class="pre">cache:</span> <span class="pre">bool</span> <span class="pre">=</span> <span class="pre">True,</span> <span class="pre">aliases:</span> <span class="pre">typing.Dict[str,</span> <span class="pre">typing.List[str]]</span> <span class="pre">=</span> <span class="pre">{},</span> <span class="pre">target_site_counter:</span> <span class="pre">typing.Optional[fastdfe.parser.TargetSiteCounter]</span> <span class="pre">=</span> <span class="pre">None,</span> <span class="pre">subsample_mode:</span> <span class="pre">typing.Literal['random',</span> <span class="pre">'probabilistic']</span> <span class="pre">=</span> <span class="pre">'probabilistic'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.Parser" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">MultiHandler</span></code></p>
<p>Parse site-frequency spectra from VCF files.</p>
<p>By default, the parser looks at the <code class="docutils literal notranslate"><span class="pre">AA</span></code> tag in the VCF file’s info field to retrieve
the correct polarization. Sites for which this tag is not well-defined are by default
ignored (see <code class="docutils literal notranslate"><span class="pre">skip_non_polarized</span></code>).</p>
<p>This class also offers on-the-fly annotation of the VCF sites such as site degeneracy and
ancestral allele state. This is done by providing a list of annotations to the parser which are
applied in the order they are provided.</p>
<p>The parser also allows to filter sites based on site properties which is done by
passing a list of filtrations. By default, we filter out poly-allelic sites as sites are assumed to be
at most bi-allelic.</p>
<p>In addition, the parser allows to stratify the SFS by providing a list of stratifications. This is useful
to obtain the SFS for different types of sites for which we can jointly infer the DFEs using
<a class="reference internal" href="inference.html#fastdfe.joint_inference.JointInference" title="fastdfe.joint_inference.JointInference"><code class="xref py py-class docutils literal notranslate"><span class="pre">JointInference</span></code></a>.</p>
<p>To correctly determine the number of target sites when parsing a VCF file that does not contain monomorphic sites,
we can use a <a class="reference internal" href="#fastdfe.parser.TargetSiteCounter" title="fastdfe.parser.TargetSiteCounter"><code class="xref py py-class docutils literal notranslate"><span class="pre">TargetSiteCounter</span></code></a>. This class is used in conjunction with the parser and
samples sites from the given FASTA file that are found in between variants on the same contig that were parsed
in the VCF.</p>
<p>Note that we assume the sites in the VCF file to be sorted by position in ascending order (per contig).</p>
<p>Example usage:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fastdfe</span> <span class="k">as</span> <span class="nn">fd</span>

<span class="c1"># Parse selected and neutral SFS from human chromosome 1.</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">fd</span><span class="o">.</span><span class="n">Parser</span><span class="p">(</span>
    <span class="n">vcf</span><span class="o">=</span><span class="s2">&quot;https://ngs.sanger.ac.uk/production/hgdp/hgdp_wgs.20190516/&quot;</span>
        <span class="s2">&quot;hgdp_wgs.20190516.full.chr1.vcf.gz&quot;</span><span class="p">,</span>
    <span class="n">fasta</span><span class="o">=</span><span class="s2">&quot;http://ftp.ensembl.org/pub/release-109/fasta/homo_sapiens/&quot;</span>
          <span class="s2">&quot;dna/Homo_sapiens.GRCh38.dna.chromosome.1.fa.gz&quot;</span><span class="p">,</span>
    <span class="n">gff</span><span class="o">=</span><span class="s2">&quot;http://ftp.ensembl.org/pub/release-109/gff3/homo_sapiens/&quot;</span>
        <span class="s2">&quot;Homo_sapiens.GRCh38.109.chromosome.1.gff3.gz&quot;</span><span class="p">,</span>
    <span class="n">aliases</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">chr1</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">]),</span>  <span class="c1"># mapping for contig names</span>
    <span class="n">n</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>  <span class="c1"># SFS sample size</span>
    <span class="c1"># we use a target site counter to infer the number of target sites.</span>
    <span class="n">target_site_counter</span><span class="o">=</span><span class="n">fd</span><span class="o">.</span><span class="n">TargetSiteCounter</span><span class="p">(</span>
        <span class="n">n_samples</span><span class="o">=</span><span class="mi">1000000</span><span class="p">,</span>
        <span class="c1"># determine number of target sites by looking at total length of coding sequences</span>
        <span class="n">n_target_sites</span><span class="o">=</span><span class="n">fd</span><span class="o">.</span><span class="n">Annotation</span><span class="o">.</span><span class="n">count_target_sites</span><span class="p">(</span>
            <span class="s2">&quot;http://ftp.ensembl.org/pub/release-109/gff3/homo_sapiens/&quot;</span>
            <span class="s2">&quot;Homo_sapiens.GRCh38.109.chromosome.1.gff3.gz&quot;</span>
        <span class="p">)[</span><span class="s1">&#39;1&#39;</span><span class="p">]</span>
    <span class="p">),</span>
    <span class="c1"># add degeneracy annotation for sites</span>
    <span class="n">annotations</span><span class="o">=</span><span class="p">[</span>
        <span class="n">fd</span><span class="o">.</span><span class="n">DegeneracyAnnotation</span><span class="p">()</span>
    <span class="p">],</span>
    <span class="n">filtrations</span><span class="o">=</span><span class="p">[</span>
        <span class="c1"># exclude non-SNPs as we infer monomorphic sites with target site counter</span>
        <span class="n">fd</span><span class="o">.</span><span class="n">SNPFiltration</span><span class="p">(),</span>
        <span class="c1"># filter out sites not in coding sequences</span>
        <span class="n">fd</span><span class="o">.</span><span class="n">CodingSequenceFiltration</span><span class="p">()</span>
    <span class="p">],</span>
    <span class="c1"># stratify by 4-fold/0-fold degeneracy</span>
    <span class="n">stratifications</span><span class="o">=</span><span class="p">[</span><span class="n">fd</span><span class="o">.</span><span class="n">DegeneracyStratification</span><span class="p">()],</span>
    <span class="n">info_ancestral</span><span class="o">=</span><span class="s1">&#39;AA_ensembl&#39;</span>
<span class="p">)</span>

<span class="n">sfs</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">parse</span><span class="p">()</span>

<span class="n">sfs</span><span class="o">.</span><span class="n">plot</span><span class="p">()</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.Parser.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">vcf:</span> <span class="pre">str,</span> <span class="pre">n:</span> <span class="pre">int,</span> <span class="pre">gff:</span> <span class="pre">typing.Optional[str]</span> <span class="pre">=</span> <span class="pre">None,</span> <span class="pre">fasta:</span> <span class="pre">typing.Optional[str]</span> <span class="pre">=</span> <span class="pre">None,</span> <span class="pre">info_ancestral:</span> <span class="pre">str</span> <span class="pre">=</span> <span class="pre">'AA',</span> <span class="pre">skip_non_polarized:</span> <span class="pre">bool</span> <span class="pre">=</span> <span class="pre">True,</span> <span class="pre">stratifications:</span> <span class="pre">typing.List[fastdfe.parser.Stratification]</span> <span class="pre">=</span> <span class="pre">[],</span> <span class="pre">annotations:</span> <span class="pre">typing.List[fastdfe.annotation.Annotation]</span> <span class="pre">=</span> <span class="pre">[],</span> <span class="pre">filtrations:</span> <span class="pre">typing.List[fastdfe.filtration.Filtration]</span> <span class="pre">=</span> <span class="pre">[&lt;fastdfe.filtration.PolyAllelicFiltration</span> <span class="pre">object&gt;],</span> <span class="pre">include_samples:</span> <span class="pre">typing.Optional[typing.List[str]]</span> <span class="pre">=</span> <span class="pre">None,</span> <span class="pre">exclude_samples:</span> <span class="pre">typing.Optional[typing.List[str]]</span> <span class="pre">=</span> <span class="pre">None,</span> <span class="pre">max_sites:</span> <span class="pre">int</span> <span class="pre">=</span> <span class="pre">inf,</span> <span class="pre">seed:</span> <span class="pre">int</span> <span class="pre">|</span> <span class="pre">None</span> <span class="pre">=</span> <span class="pre">0,</span> <span class="pre">cache:</span> <span class="pre">bool</span> <span class="pre">=</span> <span class="pre">True,</span> <span class="pre">aliases:</span> <span class="pre">typing.Dict[str,</span> <span class="pre">typing.List[str]]</span> <span class="pre">=</span> <span class="pre">{},</span> <span class="pre">target_site_counter:</span> <span class="pre">typing.Optional[fastdfe.parser.TargetSiteCounter]</span> <span class="pre">=</span> <span class="pre">None,</span> <span class="pre">subsample_mode:</span> <span class="pre">typing.Literal['random',</span> <span class="pre">'probabilistic']</span> <span class="pre">=</span> <span class="pre">'probabilistic'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.Parser.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the parser.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vcf</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – The path to the VCF file, can be gzipped or a URL.</p></li>
<li><p><strong>gff</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – The path to the GFF file, possibly gzipped or a URL. This file is optional and depends on
the stratifications, annotations and filtrations that are used.</p></li>
<li><p><strong>fasta</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – The path to the FASTA file, possibly gzipped or a URL. This file is optional and depends on
the annotations and filtrations that are used.</p></li>
<li><p><strong>n</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – The size of the resulting SFS. We down-sample to this number by drawing without replacement from
the set of all available genotypes per site. Sites with fewer than <code class="docutils literal notranslate"><span class="pre">n</span></code> genotypes are skipped.</p></li>
<li><p><strong>info_ancestral</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – The tag in the INFO field that contains ancestral allele information. Consider using
an ancestral allele annotation if this information is not available yet.</p></li>
<li><p><strong>skip_non_polarized</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to skip poly-morphic sites that are not polarized, i.e., without a valid
info tag providing the ancestral allele. If <code class="docutils literal notranslate"><span class="pre">False</span></code>, we use the reference allele as ancestral allele (
only recommended if working with folded spectra).</p></li>
<li><p><strong>stratifications</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<a class="reference internal" href="#fastdfe.parser.Stratification" title="fastdfe.parser.Stratification"><code class="xref py py-class docutils literal notranslate"><span class="pre">Stratification</span></code></a>]) – List of stratifications to use.</p></li>
<li><p><strong>annotations</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<a class="reference internal" href="annotation.html#fastdfe.annotation.Annotation" title="fastdfe.annotation.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a>]) – List of annotations to use.</p></li>
<li><p><strong>filtrations</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<a class="reference internal" href="filtration.html#fastdfe.filtration.Filtration" title="fastdfe.filtration.Filtration"><code class="xref py py-class docutils literal notranslate"><span class="pre">Filtration</span></code></a>]) – List of filtrations to use.</p></li>
<li><p><strong>include_samples</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – List of sample names to consider when determining the SFS. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, all samples
are used. Note that this restriction does not apply to the annotations and filtrations.</p></li>
<li><p><strong>exclude_samples</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – List of sample names to exclude when determining the SFS. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, no samples
are excluded. Note that this restriction does not apply to the annotations and filtrations.</p></li>
<li><p><strong>max_sites</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Maximum number of sites to parse from the VCF file.</p></li>
<li><p><strong>seed</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – Seed for the random number generator. Use <code class="docutils literal notranslate"><span class="pre">None</span></code> for no seed.</p></li>
<li><p><strong>cache</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to cache files downloaded from URLs.</p></li>
<li><p><strong>aliases</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – Dictionary of aliases for the contigs in the VCF file, e.g. <code class="docutils literal notranslate"><span class="pre">{'chr1':</span> <span class="pre">['1']}</span></code>.
This is used to match the contig names in the VCF file with the contig names in the FASTA file and GFF file.</p></li>
<li><p><strong>target_site_counter</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference internal" href="#fastdfe.parser.TargetSiteCounter" title="fastdfe.parser.TargetSiteCounter"><code class="xref py py-class docutils literal notranslate"><span class="pre">TargetSiteCounter</span></code></a>]) – The target site counter. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, we do not sample target sites.</p></li>
<li><p><strong>subsample_mode</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code>[<code class="docutils literal notranslate"><span class="pre">'random'</span></code>, <code class="docutils literal notranslate"><span class="pre">'probabilistic'</span></code>]) – The subsampling mode. For <code class="docutils literal notranslate"><span class="pre">random</span></code>, we draw once without replacement from the set of all
available genotypes per site. For <code class="docutils literal notranslate"><span class="pre">probabilistic</span></code>, we add up the hypergeometric distribution for all
sites. This will produce a smoother SFS, especially when a small number of sites is considered.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.Parser.target_site_counter">
<span class="sig-name descname"><span class="pre">target_site_counter</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code><span class="pre">[</span><a class="reference internal" href="#fastdfe.parser.TargetSiteCounter" title="fastdfe.parser.TargetSiteCounter"><code class="xref py py-class docutils literal notranslate"><span class="pre">TargetSiteCounter</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#fastdfe.parser.Parser.target_site_counter" title="Permalink to this definition"></a></dt>
<dd><p>The target site counter</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.Parser.n">
<span class="sig-name descname"><span class="pre">n</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></em><a class="headerlink" href="#fastdfe.parser.Parser.n" title="Permalink to this definition"></a></dt>
<dd><p>The number of individuals in the sample</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.Parser.include_samples">
<span class="sig-name descname"><span class="pre">include_samples</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code><span class="pre">[</span><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code><span class="pre">[</span><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code><span class="pre">]]</span></em><a class="headerlink" href="#fastdfe.parser.Parser.include_samples" title="Permalink to this definition"></a></dt>
<dd><p>The list of samples to include</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.Parser.exclude_samples">
<span class="sig-name descname"><span class="pre">exclude_samples</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code><span class="pre">[</span><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code><span class="pre">[</span><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code><span class="pre">]]</span></em><a class="headerlink" href="#fastdfe.parser.Parser.exclude_samples" title="Permalink to this definition"></a></dt>
<dd><p>The list of samples to exclude</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.Parser.skip_non_polarized">
<span class="sig-name descname"><span class="pre">skip_non_polarized</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></em><a class="headerlink" href="#fastdfe.parser.Parser.skip_non_polarized" title="Permalink to this definition"></a></dt>
<dd><p>Whether to skip sites that are not polarized, i.e., without a valid info tag providing the ancestral allele</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.Parser.stratifications">
<span class="sig-name descname"><span class="pre">stratifications</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code><span class="pre">[</span><a class="reference internal" href="#fastdfe.parser.Stratification" title="fastdfe.parser.Stratification"><code class="xref py py-class docutils literal notranslate"><span class="pre">Stratification</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#fastdfe.parser.Parser.stratifications" title="Permalink to this definition"></a></dt>
<dd><p>List of stratifications to use</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.Parser.annotations">
<span class="sig-name descname"><span class="pre">annotations</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code><span class="pre">[</span><a class="reference internal" href="annotation.html#fastdfe.annotation.Annotation" title="fastdfe.annotation.Annotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Annotation</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#fastdfe.parser.Parser.annotations" title="Permalink to this definition"></a></dt>
<dd><p>List of annotations to use</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.Parser.filtrations">
<span class="sig-name descname"><span class="pre">filtrations</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code><span class="pre">[</span><a class="reference internal" href="filtration.html#fastdfe.filtration.Filtration" title="fastdfe.filtration.Filtration"><code class="xref py py-class docutils literal notranslate"><span class="pre">Filtration</span></code></a><span class="pre">]</span></em><a class="headerlink" href="#fastdfe.parser.Parser.filtrations" title="Permalink to this definition"></a></dt>
<dd><p>List of filtrations to use</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.Parser.n_skipped">
<span class="sig-name descname"><span class="pre">n_skipped</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></em><a class="headerlink" href="#fastdfe.parser.Parser.n_skipped" title="Permalink to this definition"></a></dt>
<dd><p>The number of sites that were skipped for various reasons</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.Parser.n_no_ancestral">
<span class="sig-name descname"><span class="pre">n_no_ancestral</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></em><a class="headerlink" href="#fastdfe.parser.Parser.n_no_ancestral" title="Permalink to this definition"></a></dt>
<dd><p>The number of sites that were skipped because they had no valid ancestral allele</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.Parser.sfs">
<span class="sig-name descname"><span class="pre">sfs</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code><span class="pre">[</span><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code><span class="pre">,</span> <code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code><span class="pre">]</span></em><a class="headerlink" href="#fastdfe.parser.Parser.sfs" title="Permalink to this definition"></a></dt>
<dd><p>Dictionary of SFS indexed by joint type</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="fastdfe.parser.Parser.subsample_mode">
<span class="sig-name descname"><span class="pre">subsample_mode</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code><span class="pre">[</span><code class="docutils literal notranslate"><span class="pre">'random'</span></code><span class="pre">,</span> <code class="docutils literal notranslate"><span class="pre">'probabilistic'</span></code><span class="pre">]</span></em><a class="headerlink" href="#fastdfe.parser.Parser.subsample_mode" title="Permalink to this definition"></a></dt>
<dd><p>The subsampling mode</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fastdfe.parser.Parser.parse">
<span class="sig-name descname"><span class="pre">parse</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fastdfe.parser.Parser.parse" title="Permalink to this definition"></a></dt>
<dd><p>Parse the VCF file.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="spectra.html#fastdfe.spectrum.Spectra" title="fastdfe.spectrum.Spectra"><code class="xref py py-class docutils literal notranslate"><span class="pre">Spectra</span></code></a></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The spectra for the different stratifications</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="spectra.html" class="btn btn-neutral float-left" title="Spectra class" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="annotation.html" class="btn btn-neutral float-right" title="VCF Annotation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Janek Sendrowski.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>