<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Introduction &mdash; fastDFE 1.1.3 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" type="text/css" />
      <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Installation" href="Python/installation.html" />
    <link rel="prev" title="Welcome to fastDFE’s documentation" href="../index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            fastDFE
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Introduction</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#motivation">Motivation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#how-it-works">How it works</a></li>
<li class="toctree-l2"><a class="reference internal" href="#features">Features</a></li>
<li class="toctree-l2"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Python Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Python/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="Python/quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="Python/inference.html">DFE Inference</a></li>
<li class="toctree-l1"><a class="reference internal" href="Python/config.html">Configuration files</a></li>
<li class="toctree-l1"><a class="reference internal" href="Python/parametrizations.html">DFE parametrizations</a></li>
<li class="toctree-l1"><a class="reference internal" href="Python/spectra.html">Working with SFS</a></li>
<li class="toctree-l1"><a class="reference internal" href="Python/parser.html">VCF Parsing</a></li>
<li class="toctree-l1"><a class="reference internal" href="Python/miscellaneous.html">Miscellaneous</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">R Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="R/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="R/quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="R/inference.html">DFE Inference</a></li>
<li class="toctree-l1"><a class="reference internal" href="R/config.html">Configuration files</a></li>
<li class="toctree-l1"><a class="reference internal" href="R/parametrizations.html">DFE parametrizations</a></li>
<li class="toctree-l1"><a class="reference internal" href="R/spectra.html">Working with SFS</a></li>
<li class="toctree-l1"><a class="reference internal" href="R/parser.html">VCF Parsing</a></li>
<li class="toctree-l1"><a class="reference internal" href="R/miscellaneous.html">Miscellaneous</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../modules/inference.html">Inference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/config.html">Configuration class</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/parametrization.html">DFE Parametrizations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/spectrum.html">Spectrum class</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/spectra.html">Spectra class</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/parser.html">VCF Parsing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/annotation.html">VCF Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules/filtration.html">VCF Filtration</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">fastDFE</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Introduction</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/reference/introduction.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="introduction">
<span id="id1"></span><h1>Introduction<a class="headerlink" href="#introduction" title="Permalink to this heading"></a></h1>
<p>fastDFE is a software package designed for estimating the distribution of fitness effects (DFE) from site-frequency spectra (SFS). Building upon the foundation laid by polyDFE <span id="id2">[<a class="reference internal" href="#id6" title="Paula Tataru, Maeva Mollion, Sylvain Glémin, and Thomas Bataillon. Inference of distribution of fitness effects and proportion of adaptive substitutions from polymorphism data. Genetics, 207:genetics.300323.2017, 09 2017. doi:10.1534/genetics.117.300323.">4</a>]</span>, fastDFE addresses the limitations of its predecessor by providing a faster, more flexible, and user-friendly approach to DFE inference.</p>
<p>fastDFE is implemented in Python but also offers compatibility with R through the reticulate package, ensuring a smooth integration with existing workflows. In the R wrapper, visualizations have been reimplemented for a consistent user experience. The package is thoroughly documented and tested, ensuring its reliability and ease of use for researchers in the field.</p>
<section id="motivation">
<h2>Motivation<a class="headerlink" href="#motivation" title="Permalink to this heading"></a></h2>
<p>The DFE is instrumental in understanding how natural selection shapes genetic variation. SFS-based methods for estimating the DFE condense population genetic variation by quantifying the number of alleles at specific frequencies, discarding details about which sites exhibited which frequencies. As a result, we obtain one DFE for all variants used. This approach provides limited information when considering a single DFE, but its utility significantly increases when comparing DFEs across multiple species, populations, or genomic regions. fastDFE improves upon the groundwork laid by polyDFE, overcoming its constraints such as extensive computational time and limited scalability <span id="id3">[<a class="reference internal" href="#id7" title="Paula Tataru and Thomas Bataillon. polyDFEv2.0: testing for invariance of the distribution of fitness effects within and across species. Bioinformatics, 35(16):2868-2869, 01 2019. doi:10.1093/bioinformatics/bty1060.">3</a>, <a class="reference internal" href="#id6" title="Paula Tataru, Maeva Mollion, Sylvain Glémin, and Thomas Bataillon. Inference of distribution of fitness effects and proportion of adaptive substitutions from polymorphism data. Genetics, 207:genetics.300323.2017, 09 2017. doi:10.1534/genetics.117.300323.">4</a>]</span>. It is specifically designed to facilitate and encourage the joint inference of multiple DFEs at once. Given the importance of consistency when comparing DFEs from different datasets, it’s essential to derive the SFS in a similar manner to ensure results are directly comparable. To aid this process, fastDFE includes a VCF parser, enabling the extraction of the necessary SFS input data from raw VCF files.</p>
</section>
<section id="how-it-works">
<h2>How it works<a class="headerlink" href="#how-it-works" title="Permalink to this heading"></a></h2>
<p>As it precursors, fastDFE infers the DFE by contrasting the site-frequency spectrum (SFS) of putatively neutral (synonymous) and selected (nonsynonymous) sites <span id="id4">[<a class="reference internal" href="#id9" title="Nicolas Galtier. Adaptive protein evolution in animals and the effective population size hypothesis. PLoS genetics, 12:e1005774, 01 2016. doi:10.1371/journal.pgen.1005774.">1</a>, <a class="reference internal" href="#id8" title="Peter Keightley and Adam Eyre-Walker. Joint inference of the distribution of fitness effects of deleterious mutations and population demography based on nucleotide polymorphism frequencies. Genetics, 177:2251-61, 01 2008. doi:10.1534/genetics.107.080663.">2</a>, <a class="reference internal" href="#id6" title="Paula Tataru, Maeva Mollion, Sylvain Glémin, and Thomas Bataillon. Inference of distribution of fitness effects and proportion of adaptive substitutions from polymorphism data. Genetics, 207:genetics.300323.2017, 09 2017. doi:10.1534/genetics.117.300323.">4</a>]</span>. fastDFE builds upon the model of polyDFE, introducing several improvements: The impact of demography is corrected for by introducing nuisance parameters which are fit so as to re-scale the observed neutral SFS to the standard Kingman SFS. Assuming that demography affects the neutral and selected SFS alike, we apply the same re-scaling to the observed selected SFS. Unlike polyDFE, demography is inferred solely from the synonymous sites, which should provide sufficient information for reasonably large sample sizes, and has the benefit to speed up calculations. To infer selection from the (demography-corrected) SFS, we also make use of the expected allele frequency sojourn times obtained from Poisson random field (PRF) theory. Obtaining the modelled (selected) SFS requires integrating over the DFE in question. In fastDFE, this integral is discretized, precomputed and cached, leading to significant speed enhancements over polyDFE while producing very similar results. Discontinuities and asymptotes are furthermore handled properly by parametrizing the DFE by means of its cumulative distribution function, which provides more reliable estimates. Ultimately, the most likely DFE is obtained using maximum likelihood estimation (MLE).</p>
</section>
<section id="features">
<h2>Features<a class="headerlink" href="#features" title="Permalink to this heading"></a></h2>
<p>fastDFE offers a range of features designed to facilitate DFE inference.</p>
<p><strong>Workflow</strong>: streamlining the overall data preparation and inference process:</p>
<ul class="simple">
<li><p>Built-in VCF-to-SFS parser, with support for versatile stratification, site annotation, and filtering</p></li>
<li><p>Site-degeneracy annotation</p></li>
<li><p>Ancestral-allele annotation with outgroups</p></li>
<li><p>Utilities to determine the number of mutational target sites when monomorphic sites are not present in the provided VCF file</p></li>
<li><p>Support for configuration files, facilitating reproducibility</p></li>
<li><p>Serialization of objects and result summaries</p></li>
<li><p>Object-oriented and customizable design</p></li>
</ul>
<p><strong>Modeling</strong>: robust utilities for fitting models to data and estimating DFE parameters:</p>
<ul class="simple">
<li><p>Joint inference of DFE across different types which share parameters</p></li>
<li><p>Inclusion of covariates and likelihood ratio tests (LRTs) to assess their significance</p></li>
<li><p>Parametric bootstrapping</p></li>
<li><p>Nested model comparison with LRTs</p></li>
<li><p>Customizable DFE parametrizations</p></li>
<li><p>Estimation of <span class="math notranslate nohighlight">\(\alpha\)</span>, the proportion of beneficial substitutions</p></li>
</ul>
<p><strong>Visualization</strong>: encompassing a variety of visualization utilities:</p>
<ul class="simple">
<li><p>Visualization of DFE, nested model p-values, inferred parameters, and their confidence intervals and more</p></li>
<li><p>Support for both Python and R</p></li>
</ul>
</section>
<section id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this heading"></a></h2>
<div class="docutils container" id="id5">
<dl class="citation">
<dt class="label" id="id9"><span class="brackets"><a class="fn-backref" href="#id4">1</a></span></dt>
<dd><p>Nicolas Galtier. Adaptive protein evolution in animals and the effective population size hypothesis. <em>PLoS genetics</em>, 12:e1005774, 01 2016. <a class="reference external" href="https://doi.org/10.1371/journal.pgen.1005774">doi:10.1371/journal.pgen.1005774</a>.</p>
</dd>
<dt class="label" id="id8"><span class="brackets"><a class="fn-backref" href="#id4">2</a></span></dt>
<dd><p>Peter Keightley and Adam Eyre-Walker. Joint inference of the distribution of fitness effects of deleterious mutations and population demography based on nucleotide polymorphism frequencies. <em>Genetics</em>, 177:2251–61, 01 2008. <a class="reference external" href="https://doi.org/10.1534/genetics.107.080663">doi:10.1534/genetics.107.080663</a>.</p>
</dd>
<dt class="label" id="id7"><span class="brackets"><a class="fn-backref" href="#id3">3</a></span></dt>
<dd><p>Paula Tataru and Thomas Bataillon. polyDFEv2.0: testing for invariance of the distribution of fitness effects within and across species. <em>Bioinformatics</em>, 35(16):2868–2869, 01 2019. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bty1060">doi:10.1093/bioinformatics/bty1060</a>.</p>
</dd>
<dt class="label" id="id6"><span class="brackets">4</span><span class="fn-backref">(<a href="#id2">1</a>,<a href="#id3">2</a>,<a href="#id4">3</a>)</span></dt>
<dd><p>Paula Tataru, Maeva Mollion, Sylvain Glémin, and Thomas Bataillon. Inference of distribution of fitness effects and proportion of adaptive substitutions from polymorphism data. <em>Genetics</em>, 207:genetics.300323.2017, 09 2017. <a class="reference external" href="https://doi.org/10.1534/genetics.117.300323">doi:10.1534/genetics.117.300323</a>.</p>
</dd>
</dl>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../index.html" class="btn btn-neutral float-left" title="Welcome to fastDFE’s documentation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Python/installation.html" class="btn btn-neutral float-right" title="Installation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Janek Sendrowski.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>